File: src/python/laboneq/compiler/seqc/output_mute.py
# Copyright 2024 Zurich Instruments AG
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

import math
from typing import TYPE_CHECKING
from laboneq.core.utilities.pulse_sampler import length_to_samples
from laboneq.core.exceptions import LabOneQException


if TYPE_CHECKING:
    from laboneq.compiler.seqc.seqc_tracker import SeqCTracker
    from laboneq.compiler.common.device_type import DeviceType


def ceil(value: int, grid: int):
    return value + (-value) % grid


class OutputMute:
    """Mute played samples.

    Arguments:
        device_type: Device type
            Given device must support output mute.
        generator: SeqC generator
        duration_min: Minimum duration for mute to be fully engaged.
            Therefore the minimum mute duration is: engage time + `duration_min` + disengage time
    """

    def __init__(
        self,
        device_type: DeviceType,
        generator: SeqCTracker,
        duration_min: float,
    ):
        assert device_type.supports_output_mute, (
            f"Device {device_type.name.upper()} does not support output mute."
        )
        self._device_type = device_type
        self._generator = generator

        # The minimum time for the muting required by the instrument
        device_duration_min = (
            # latency of just turning on and off the blanking...
            self._device_type.output_mute_engage_delay
            - self._device_type.output_mute_disengage_delay
            # ... plus a minimal playZero
            + self._device_type.min_play_wave / self._device_type.sampling_rate
        )
        if duration_min <= device_duration_min:
            msg = f"Output mute duration must be larger than {device_duration_min} s."
            raise LabOneQException(msg)
        samples_min = length_to_samples(
            duration_min,
            self._device_type.sampling_rate,
        )
        self._samples_min = (
            math.ceil(samples_min / self._device_type.sample_multiple)
            * self._device_type.sample_multiple
        )
        delay_engage = length_to_samples(
            self._device_type.output_mute_engage_delay,
            self._device_type.sampling_rate,
        )
        self.delay_engage = ceil(delay_engage, self._device_type.sample_multiple)
        delay_disengage = length_to_samples(
            -self._device_type.output_mute_disengage_delay,
            self._device_type.sampling_rate,
        )
        self.delay_disengage = ceil(delay_disengage, self._device_type.sample_multiple)

    @property
    def samples_min(self) -> int:
        return self._samples_min

    def can_mute(self, samples: int) -> bool:
        return samples >= self.samples_min

    def mute_samples(self, samples: int):
        """Mute samples to be played.

        If the length of samples exceeds `samples_min`, muting is applied,
        otherwise no action is done.
        """
        if self.can_mute(samples):
            self._generator.current_loop_stack_generator().add_play_zero_statement(
                self.delay_engage,
                self._device_type,
                self._generator.deferred_function_calls,
            )
            self._generator.add_function_call_statement(
                "setTrigger", [1], deferred=True
            )
            self._generator.current_loop_stack_generator().add_play_zero_statement(
                samples - self.delay_engage - self.delay_disengage,
                self._device_type,
                self._generator.deferred_function_calls,
            )
            self._generator.add_function_call_statement(
                "setTrigger", [0], deferred=True
            )
            self._generator.current_loop_stack_generator().add_play_zero_statement(
                self.delay_disengage,
                self._device_type,
                self._generator.deferred_function_calls,
            )

File: src/rust/seqc-tracker/src/output_mute.rs
// Copyright 2025 Zurich Instruments AG
// SPDX-License-Identifier: Apache-2.0

use crate::{Result, Samples};
use anyhow::anyhow;
use codegenerator::{ir::compilation_job::DeviceKind, tinysample::length_to_samples};
use std::str::FromStr;

fn ceil(value: i64, grid: u16) -> Samples {
    assert!(value > 0, "Value must be greater than 0");
    let grid = grid as i64;
    // In Rust, the modulo operator works differently for negative numbers
    // compared to Python. In Python, -1 % 3 is 2, while in Rust it is -1.
    // Python code: value + (-value) % grid
    let mut vg = -value % grid;
    if vg < 0 {
        vg += grid;
    }
    (value + vg) as Samples
}

/// Mute played samples.
///
/// # Arguments:
///
/// * device_type: Device type
///   Given device must support output mute.
/// * duration_min: Minimum duration for mute to be fully engaged.
///   Therefore the minimum mute duration is: engage time + `duration_min` + disengage time
///
pub struct OutputMute {
    pub samples_min: Samples,
    pub delay_engage: Samples,
    pub delay_disengage: Samples,
}

impl OutputMute {
    pub fn new(device_type: &str, duration_min: f64) -> Result<Self> {
        let device_kind = DeviceKind::from_str(device_type)?;

        let device_traits = device_kind.traits();
        if !device_traits.supports_output_mute {
            return Err(anyhow!(
                "Unsupported device type: {0}. Supported types are: SHFQA, SHFSG and SHFQC.",
                device_type.to_uppercase()
            )
            .into());
        }

        // The minimum time for the muting required by the instrument
        let device_duration_min = device_traits.output_mute_engage_delay
            - device_traits.output_mute_disengage_delay  // latency of just turning on and off the blanking...
            + device_traits.min_play_wave as f64 / device_traits.sampling_rate; // ... plus a minimal playZero
        if duration_min <= device_duration_min {
            return Err(anyhow!(
                "Output mute duration must be larger than {} s.",
                device_duration_min
            )
            .into());
        }
        let samples_min = length_to_samples(duration_min, device_traits.sampling_rate);
        let samples_min = ((samples_min as f64 / device_traits.sample_multiple as f64).ceil()
            * device_traits.sample_multiple as f64) as Samples;
        let delay_engage = length_to_samples(
            device_traits.output_mute_engage_delay,
            device_traits.sampling_rate,
        );
        let delay_engage = ceil(delay_engage, device_traits.sample_multiple) as Samples;
        let delay_disengage = length_to_samples(
            -device_traits.output_mute_disengage_delay,
            device_traits.sampling_rate,
        );
        let delay_disengage = ceil(delay_disengage, device_traits.sample_multiple);

        Ok(OutputMute {
            samples_min,
            delay_engage,
            delay_disengage,
        })
    }
}
